README.md
Cloudinary
==========

Cloudinary is a cloud service that offers a solution to a web application's entire image management pipeline.

Easily upload images to the cloud. Automatically perform smart image resizing, cropping and conversion without installing any complex software. Integrate Facebook or Twitter profile image extraction in a snap, in any dimension and style to match your website’s graphics requirements. Images are seamlessly delivered through a fast CDN, and much much more.

Cloudinary offers comprehensive APIs and administration capabilities and is easy to integrate with any web application, existing or new.

Cloudinary provides URL and HTTP based APIs that can be easily integrated with any Web development framework.

For Titanium, Cloudinary provides a client-side JavaScript library for simplifying the integration with Titanium supported mobile platforms (Android, iOS, BlackBerry, Tizen) even further.

## Setup

* [Sign up](https://cloudinary.com/users/register/free) for a free Cloudinary account
* [Register, download Titanium SDK, setup Titanium Studio and start a project](http://www.appcelerator.com/developers/)
* [Download](https://github.com/cloudinary/cloudinary_titanium/archive/master.zip) or [clone](https://github.com/cloudinary/cloudinary_titanium.git) this repository
* Copy the Resources/lib folder into your project (if you already have one you can merge the content)
* Copy the `cloudinary_config.js.sample` into `cloudinary_config.js` in edit it to reflect your Cloudinary cloud name.

## Usage

The Cloudinary Documentation can be found at: http://cloudinary.com/documentation

Loading the module:

    var cloudinary = require('/lib/cloudinary');

Retrieving a cloudinary URL based on your on your configuration and the given options:

    var url = cloudinary.url(public_id, options);

Uploading an image (requires a backend with Cloudinary framework):

    var signed_request = retreive_signed_request_from_backend()
    var file = Ti.Filesystem.getFile('my_file')
    function callback(result) {
        if (result.error) {
            Ti.API.error("Error: " + result.error);
        } else {
            Ti.API.info("Uploaded file with public_id: " + result.public_id);
    }
    cloudinary.uploader.upload(file, callback, signed_request);


Uploading an image with `api_secret` (Not recommended - your API secret will be exposed to users!)

    var file = Ti.Filesystem.getFile('my_file')
    function callback(result) {
        if (result.error) {
            Ti.API.error("Error: " + result.error);
        } else {
            Ti.API.info("Uploaded file with public_id: " + result.public_id);
    }
    cloudinary.uploader.upload(file, callback, {api_key: "my_api_key", api_secret: "my_api_secret");

Managing resources:

    cloudinary.uploader.destroy(public_id, callback, options)
    cloudinary.uploader.rename(from_public_id, to_public_id, callback, options)

Manipulating tags:

    cloudinary.uploader.add_tag(tag, public_ids, callback, options)
    cloudinary.uploader.remove_tag(tag, public_ids, callback, options)
    cloudinary.uploader.replace_tag(tag, public_ids, callback, options)

Advanced features (please see [documentation](http://cloudinary.com/documentation/sprite_generation) for usage specifications):

    cloudinary.uploader.explicit(public_id, callback, options)  // Refresh a Facebook/Twitter image
    cloudinary.uploader.text(text, callback, options)           // Generate a text image
    cloudinary.uploader.generate_sprite(tag, callback, options) // Generate a sprite
    cloudinary.uploader.multi(tag, callback, options)           // Create an animated GIF or a multi-page PDF document
    cloudinary.uploader.explode(public_id, callback, options)   // Explode a PDF into multiple separate images

## Additional resources

Additional resources are available at:

* [Website](http://cloudinary.com)
* [Documentation](http://cloudinary.com/documentation)
* [Knowledge Base](http://support.cloudinary.com/forums)
* [Upload API documentation](http://cloudinary.com/documentation/upload_images)

## Support

You can [open an issue through GitHub](https://github.com/cloudinary/cloudinary_titanium/issues).

Contact us [http://cloudinary.com/contact](http://cloudinary.com/contact)

Stay tuned for updates, tips and tutorials: [Blog](http://cloudinary.com/blog), [Twitter](https://twitter.com/cloudinary), [Facebook](http://www.facebook.com/Cloudinary).


## License #######################################################################

Released under the MIT license.


CHANGELOG.txt
v0.1
 - Initial Cloudinary integration with Titanium


Resources/spec/env.js
// I don't like this at all, but it beats having to call
// 'jasmine.it()' or 'jasmine.expect()' all the freaking time.
var __j = require('spec/lib/jasmine-1.2.0');
var methods = ['spyOn',
               'it',
               'xit',
               'expect',
               'runs',
               'waits',
               'waitsFor',
               'beforeEach',
               'afterEach',
               'describe',
               'xdescribe'
               ];

for (var i = 0, l = methods.length; i < l;  i++) {
    var method = methods[i];
    this[method] = __j[method];
    Ti.API.debug('require ' + method);
}
this.jasmine = __j;


Resources/spec/tests.js
exports.execute = function() {
    var jasmine = require('spec/lib/jasmine-1.2.0').jasmine;
    var reporter = require('spec/lib/reporter').TitaniumReporter;
    Ti.include('spec/env.js');

    require('spec/cloudinary_spec');
    require('spec/uploader_spec');

    jasmine.getEnv().addReporter(new reporter());
    jasmine.getEnv().execute();
};



Resources/spec/api_spec.js
(function() {

  describe("cloudinary_uploader", function() {
    var API_TIMEOUT, UPLOAD_TIMEOUT, cloudinary, find_by_attr;
    cloudinary = require('/lib/cloudinary');
    find_by_attr = function(elements, attr, value) {
      var element, _i, _len;
      for (_i = 0, _len = elements.length; _i < _len; _i++) {
        element = elements[_i];
        if (element[attr] === value) {
          return element;
        }
      }
      return void 0;
    };
    UPLOAD_TIMEOUT = 120 * 1000;
    API_TIMEOUT = 60 * 1000;
    if (!(cloudinary.config().api_secret != null)) {
      return console.log("JASMINE: **** Please setup environment for uploader test to run!");
    }
    beforeEach(function() {
      var progress, semaphore;
      cloudinary.config(true);
      semaphore = 3;
      progress = function() {
        return semaphore -= 1;
      };
      runs(function() {
        return cloudinary.uploader.destroy("api_test", function() {
          return cloudinary.uploader.destroy("api_test2", function() {
            cloudinary.uploader.upload("spec/logo.png", progress, {
              public_id: "api_test",
              tags: "api_test_tag",
              eager: [
                {
                  width: 100,
                  crop: "scale"
                }
              ]
            });
            cloudinary.uploader.upload("spec/logo.png", progress, {
              public_id: "api_test2",
              tags: "api_test_tag",
              eager: [
                {
                  width: 100,
                  crop: "scale"
                }
              ]
            });
            return cloudinary.api.delete_transformation("api_test_transformation", progress);
          });
        });
      });
      return waitsFor((function() {
        return semaphore === 0;
      }), 'setup should finish', UPLOAD_TIMEOUT);
    });
    xit("should allow listing resource_types", function() {
      var result;
      result = void 0;
      runs(function() {
        return cloudinary.api.resource_types(function(result_) {
          return result = result_;
        }, {
          timeout: API_TIMEOUT
        });
      });
      waitsFor((function() {
        return result;
      }), 'resource_types should finish', API_TIMEOUT);
      return runs(function() {
        expect(result.error).toBe(void 0);
        return expect(result.resource_types).toContain("image");
      });
    });
    xit("should allow listing resources", function() {
      var result;
      result = void 0;
      runs(function() {
        return cloudinary.api.resources(function(result_) {
          return result = result_;
        }, {
          timeout: API_TIMEOUT
        });
      });
      waitsFor((function() {
        return result;
      }), 'resources should finish', API_TIMEOUT);
      return runs(function() {
        var resource;
        expect(result.error).toBe(void 0);
        resource = find_by_attr(result.resources, "public_id", "api_test");
        expect(resource).not.toBe(void 0);
        return expect(resource.type).toEqual("upload");
      });
    });
    xit("should allow listing resources with cursor", function() {
      var result, result2;
      result = void 0;
      runs(function() {
        return cloudinary.api.resources(function(result_) {
          return result = result_;
        }, {
          timeout: API_TIMEOUT,
          max_results: 1
        });
      });
      waitsFor((function() {
        return result;
      }), 'resources should finish', API_TIMEOUT);
      result2 = void 0;
      runs(function() {
        expect(result.error).toBe(void 0);
        expect(result.resources.length).toEqual(1);
        expect(result.next_cursor).not.toBe(void 0);
        return cloudinary.api.resources(function(result_) {
          return result2 = result_;
        }, {
          timeout: API_TIMEOUT,
          max_results: 1,
          next_cursor: result.next_cursor
        });
      });
      waitsFor((function() {
        return result2;
      }), 'resources should finish', API_TIMEOUT);
      return runs(function() {
        expect(result2.error).toBe(void 0);
        expect(result2.resources.length).toEqual(1);
        expect(result2.next_cursor).not.toBe(void 0);
        expect(result2.resources[0].public_id).not.toBe(void 0);
        return expect(result.resources[0].public_id).not.toEqual(result2.resources[0].public_id);
      });
    });
    xit("should allow listing resources by type", function() {
      var result;
      result = void 0;
      runs(function() {
        return cloudinary.api.resources(function(result_) {
          return result = result_;
        }, {
          timeout: API_TIMEOUT,
          type: "upload"
        });
      });
      waitsFor((function() {
        return result;
      }), 'resources should finish', API_TIMEOUT);
      return runs(function() {
        var resource;
        expect(result.error).toBe(void 0);
        resource = find_by_attr(result.resources, "public_id", "api_test");
        expect(resource).not.toBe(void 0);
        return expect(resource.type).toEqual("upload");
      });
    });
    xit("should allow listing resources by prefix", function() {
      var result;
      result = void 0;
      runs(function() {
        return cloudinary.api.resources(function(result_) {
          return result = result_;
        }, {
          timeout: API_TIMEOUT,
          type: "upload",
          prefix: "api_test"
        });
      });
      waitsFor((function() {
        return result;
      }), 'resources should finish', API_TIMEOUT);
      return runs(function() {
        var public_ids, resource;
        expect(result.error).toBe(void 0);
        public_ids = (function() {
          var _i, _len, _ref, _results;
          _ref = result.resources;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            resource = _ref[_i];
            _results.push(resource.public_id);
          }
          return _results;
        })();
        expect(public_ids).toContain("api_test");
        return expect(public_ids).toContain("api_test2");
      });
    });
    xit("should allow listing resources by tag", function() {
      var result;
      result = void 0;
      runs(function() {
        return cloudinary.api.resources_by_tag("api_test_tag", function(result_) {
          return result = result_;
        }, {
          timeout: API_TIMEOUT,
          type: "upload",
          prefix: "api_test"
        });
      });
      waitsFor((function() {
        return result;
      }), 'resources_by_tag should finish', API_TIMEOUT);
      return runs(function() {
        var resource;
        expect(result.error).toBe(void 0);
        resource = find_by_attr(result.resources, "public_id", "api_test");
        expect(resource).not.toBe(void 0);
        return expect(resource.type).toEqual("upload");
      });
    });
    xit("should allow get resource metadata", function() {
      var resource;
      resource = void 0;
      runs(function() {
        return cloudinary.api.resource("api_test", function(result_) {
          return resource = result_;
        }, {
          timeout: API_TIMEOUT,
          type: "upload",
          prefix: "api_test"
        });
      });
      waitsFor((function() {
        return resource;
      }), 'resource should finish', API_TIMEOUT);
      return runs(function() {
        expect(resource).not.toBe(void 0);
        expect(resource.error).toBe(void 0);
        expect(resource.public_id).toEqual("api_test");
        expect(resource.bytes).toEqual(3381);
        return expect(resource.derived.length).toEqual(1);
      });
    });
    it("should allow deleting derived resource", function(done) {
      var derived_resource_id, resource, result;
      result = void 0;
      runs(function() {
        return cloudinary.uploader.upload("spec/logo.png", function(result_) {
          return result = result_;
        }, {
          timeout: UPLOAD_TIMEOUT,
          public_id: "api_test3",
          eager: [
            {
              width: 101,
              crop: "scale"
            }
          ]
        });
      });
      waitsFor((function() {
        return result;
      }), 'upload should finish', UPLOAD_TIMEOUT);
      resource = void 0;
      runs(function() {
        expect(result.error).toBe(void 0);
        return cloudinary.api.resource("api_test3", function(result_) {
          return resource = result_;
        });
      });
      waitsFor((function() {
        return resource;
      }), 'resource should finish', API_TIMEOUT);
      derived_resource_id = void 0;
      runs(function() {
        expect(resource).not.toBe(void 0);
        expect(resource.bytes).toEqual(3381);
        expect(resource.derived.length).toEqual(1);
        derived_resource_id = resource.derived[0].id;
        result = void 0;
        return cloudinary.api.delete_derived_resources(derived_resource_id, function(result_) {
          return result = result_;
        });
      });
      waitsFor((function() {
        return result;
      }), 'delete_derived_resources should finish', API_TIMEOUT);
      derived_resource_id = void 0;
      runs(function() {
        expect(result.error).toBe(void 0);
        resource = void 0;
        return cloudinary.api.resource("api_test3", function(result_) {
          return resource = result_;
        });
      });
      waitsFor((function() {
        return resource;
      }), 'resource should finish', API_TIMEOUT);
      return runs(function() {
        expect(resource).not.toBe(void 0);
        return expect(resource.derived.length).toEqual(0);
      });
    });
    xit("should allow deleting resources", function(done) {
      var result;
      result = void 0;
      runs(function() {
        return cloudinary.uploader.upload("spec/logo.png", function(result_) {
          return result = result_;
        }, {
          timeout: UPLOAD_TIMEOUT,
          public_id: "api_test3"
        });
      });
      waitsFor((function() {
        return result;
      }), 'upload should finish', UPLOAD_TIMEOUT);
      runs(function() {
        expect(result.error).toBe(void 0);
        result = void 0;
        return cloudinary.api.resource("api_test3", function(result_) {
          return result = result_;
        });
      });
      waitsFor((function() {
        return result;
      }), 'resource should finish', API_TIMEOUT);
      runs(function() {
        expect(result).not.toBe(void 0);
        result = void 0;
        return cloudinary.api.delete_resources(["apit_test", "api_test2", "api_test3"], function() {
          return cloudinary.api.resource("api_test3", function(result) {
            return result = result_;
          });
        });
      });
      waitsFor((function() {
        return result;
      }), 'resource should finish', API_TIMEOUT);
      return runs(function() {
        expect(result).not.toBe(void 0);
        return expect(result.error.http_code).toEqual(404);
      });
    });
    xit("should allow deleting resources by prefix", function(done) {
      var resource, result;
      result = void 0;
      runs(function() {
        return cloudinary.uploader.upload("spec/logo.png", function(result_) {
          return result = result_;
        }, {
          timeout: UPLOAD_TIMEOUT,
          public_id: "api_test_by_prefix"
        });
      });
      waitsFor((function() {
        return result;
      }), 'upload should finish', UPLOAD_TIMEOUT);
      resource = void 0;
      runs(function() {
        expect(result.error).toBe(void 0);
        return cloudinary.api.resource("api_test_by_prefix", function(result_) {
          return resource = result_;
        });
      });
      waitsFor((function() {
        return resource;
      }), 'resource should finish', API_TIMEOUT);
      runs(function() {
        expect(resource).not.toBe(void 0);
        result = void 0;
        return cloudinary.api.delete_resources_by_prefix("api_test_by", function(result_) {
          return result = result_;
        });
      });
      waitsFor((function() {
        return result;
      }), 'delete_resources_by_prefix should finish', API_TIMEOUT);
      runs(function() {
        expect(result.error).toBe(void 0);
        return cloudinary.api.resource("api_test_by_prefix", function(result_) {
          return result = result_;
        });
      });
      waitsFor((function() {
        return result;
      }), 'resource should finish', API_TIMEOUT);
      return runs(function() {
        expect(result.error).toBe(void 0);
        return expect(result.error.http_code).toEqual(404);
      });
    });
    return;
    it("should allow deleting resources by tags", function(done) {
      this.timeout(10000);
      return cloudinary.uploader.upload("spec/logo.png", function(r) {
        if (r.error != null) {
          return done(new Error(r.error.message));
        }
        return cloudinary.api.resource("api_test4", function(resource) {
          expect(resource).not.toEqual(void 0);
          return cloudinary.api.delete_resources_by_tag("api_test_tag_for_delete", function(rr) {
            if (rr.error != null) {
              return done(new Error(rr.error.message));
            }
            return cloudinary.api.resource("api_test4", function(result) {
              expect(result.error).not.to.be(void 0);
              expect(result.error.http_code).toEqual(404);
              return done();
            });
          });
        });
      }, {
        public_id: "api_test4",
        tags: ["api_test_tag_for_delete"]
      });
    });
    it("should allow listing tags", function(done) {
      this.timeout(10000);
      return cloudinary.api.tags(function(result) {
        if (result.error != null) {
          return done(new Error(result.error.message));
        }
        expect(result.tags).toContain("api_test_tag");
        return done();
      });
    });
    it("should allow listing tag by prefix ", function(done) {
      this.timeout(10000);
      return cloudinary.api.tags(function(result) {
        if (result.error != null) {
          return done(new Error(result.error.message));
        }
        expect(result.tags).toContain("api_test_tag");
        return done();
      }, {
        prefix: "api_test"
      });
    });
    it("should allow listing tag by prefix if not found", function(done) {
      this.timeout(10000);
      return cloudinary.api.tags(function(result) {
        if (result.error != null) {
          return done(new Error(result.error.message));
        }
        expect(result.tags).to.have.length(0);
        return done();
      }, {
        prefix: "api_test_no_such_tag"
      });
    });
    it("should allow listing transformations", function(done) {
      this.timeout(10000);
      return cloudinary.api.transformations(function(result) {
        var transformation;
        if (result.error != null) {
          return done(new Error(result.error.message));
        }
        transformation = find_by_attr(result.transformations, "name", "c_scale,w_100");
        expect(transformation).not.toEqual(void 0);
        expect(transformation.used).to.be.ok;
        return done();
      });
    });
    it("should allow getting transformation metadata", function(done) {
      this.timeout(10000);
      return cloudinary.api.transformation("c_scale,w_100", function(transformation) {
        expect(transformation).not.toEqual(void 0);
        expect(transformation.info).toEqual([
          {
            crop: "scale",
            width: 100
          }
        ]);
        return done();
      });
    });
    it("should allow getting transformation metadata by info", function(done) {
      this.timeout(10000);
      return cloudinary.api.transformation({
        crop: "scale",
        width: 100
      }, function(transformation) {
        expect(transformation).not.toEqual(void 0);
        expect(transformation.info).toEqual([
          {
            crop: "scale",
            width: 100
          }
        ]);
        return done();
      });
    });
    it("should allow updating transformation allowed_for_strict", function(done) {
      this.timeout(10000);
      return cloudinary.api.update_transformation("c_scale,w_100", {
        allowed_for_strict: true
      }, function() {
        return cloudinary.api.transformation("c_scale,w_100", function(transformation) {
          expect(transformation).not.toEqual(void 0);
          expect(transformation.allowed_for_strict).to.be.ok;
          return cloudinary.api.update_transformation("c_scale,w_100", {
            allowed_for_strict: false
          }, function() {
            return cloudinary.api.transformation("c_scale,w_100", function(transformation) {
              expect(transformation).not.toEqual(void 0);
              expect(transformation.allowed_for_strict).not.to.be.ok;
              return done();
            });
          });
        });
      });
    });
    it("should allow creating named transformation", function(done) {
      this.timeout(10000);
      return cloudinary.api.create_transformation("api_test_transformation", {
        crop: "scale",
        width: 102
      }, function() {
        return cloudinary.api.transformation("api_test_transformation", function(transformation) {
          expect(transformation).not.toEqual(void 0);
          expect(transformation.allowed_for_strict).to.be.ok;
          expect(transformation.info).toEqual([
            {
              crop: "scale",
              width: 102
            }
          ]);
          expect(transformation.used).not.to.be.ok;
          return done();
        });
      });
    });
    it("should allow unsafe update of named transformation", function(done) {
      this.timeout(10000);
      return cloudinary.api.create_transformation("api_test_transformatio3", {
        crop: "scale",
        width: 102
      }, function() {
        return cloudinary.api.update_transformation("api_test_transformation3", {
          unsafe_update: {
            crop: "scale",
            width: 103
          }
        }, function() {
          return cloudinary.api.transformation("api_test_transformation3", function(transformation) {
            expect(transformation).not.toEqual(void 0);
            expect(transformation.info).toEqual([
              {
                crop: "scale",
                width: 103
              }
            ]);
            expect(transformation.used).not.to.be.ok;
            return done();
          });
        });
      });
    });
    it("should allow deleting named transformation", function(done) {
      this.timeout(10000);
      return cloudinary.api.delete_transformation("api_test_transformation", function() {
        return cloudinary.api.transformation("api_test_transformation", function(transformation) {
          expect(transformation.error.http_code).toEqual(404);
          return done();
        });
      });
    });
    it("should allow deleting implicit transformation", function(done) {
      this.timeout(10000);
      return cloudinary.api.transformation("c_scale,w_100", function(transformation) {
        expect(transformation).not.toEqual(void 0);
        return cloudinary.api.delete_transformation("c_scale,w_100", function() {
          return cloudinary.api.transformation("c_scale,w_100", function(transformation) {
            expect(transformation.error.http_code).toEqual(404);
            return done();
          });
        });
      });
    });
    return it("should support the usage API call", function(done) {
      this.timeout(10000);
      return cloudinary.api.usage(function(usage) {
        expect(usage.last_update).not.toEqual(null);
        return done();
      });
    });
  });

}).call(this);


Resources/lib/cloudinary.js
var _ = require('./underscore');

exports.config = require("./cloudinary/config");
exports.utils = require("./cloudinary/utils");
exports.uploader = require("./cloudinary/uploader");
exports.url = function(public_id, options) {
  options = _.extend({}, options);
  return exports.utils.url(public_id, options);    
};    
exports.image = function(source, options) {
  options = options || {};
  source = exports.utils.url(source, options);
  if ("html_width" in options) options["width"] = exports.utils.option_consume(options, "html_width")
  if ("html_height" in options) options["height"] = exports.utils.option_consume(options, "html_height")
  return "<img src='" + source + "' " + exports.utils.html_attrs(options) + "/>";
}
exports.CF_SHARED_CDN = exports.utils.CF_SHARED_CDN;
exports.AKAMAI_SHARED_CDN = exports.utils.AKAMAI_SHARED_CDN;
exports.SHARED_CDN = exports.utils.SHARED_CDN;


Resources/lib/cloudinary/api.js
(function() {
  var call_api, config, only, transformation_string, utils, _,
    __slice = [].slice;

  _ = require("../underscore");

  utils = require("./utils");

  config = require("./config");

  exports.usage = function(callback, options) {
    if (options == null) {
      options = {};
    }
    return call_api("get", ["usage"], {}, callback, options);
  };

  exports.resource_types = function(callback, options) {
    if (options == null) {
      options = {};
    }
    return call_api("get", ["resources"], {}, callback, options);
  };

  exports.resources = function(callback, options) {
    var resource_type, type, uri, _ref;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    type = options["type"];
    uri = ["resources", resource_type];
    if (type != null) {
      uri.push(type);
    }
    return call_api("get", uri, only(options, "next_cursor", "max_results", "prefix"), callback, options);
  };

  exports.resources_by_tag = function(tag, callback, options) {
    var resource_type, uri, _ref;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    uri = ["resources", resource_type, "tags", tag];
    return call_api("get", uri, only(options, "next_cursor", "max_results"), callback, options);
  };

  exports.resource = function(public_id, callback, options) {
    var resource_type, type, uri, _ref, _ref1;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    type = (_ref1 = options["type"]) != null ? _ref1 : "upload";
    uri = ["resources", resource_type, type, public_id];
    return call_api("get", uri, only(options, "exif", "colors", "faces", "image_metadata", "pages", "max_results"), callback, options);
  };

  exports.delete_resources = function(public_ids, callback, options) {
    var resource_type, type, uri, _ref, _ref1;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    type = (_ref1 = options["type"]) != null ? _ref1 : "upload";
    uri = ["resources", resource_type, type];
    return call_api("delete", uri, _.extend({
      "public_ids[]": public_ids
    }, only(options, "keep_original")), callback, options);
  };

  exports.delete_resources_by_prefix = function(prefix, callback, options) {
    var resource_type, type, uri, _ref, _ref1;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    type = (_ref1 = options["type"]) != null ? _ref1 : "upload";
    uri = ["resources", resource_type, type];
    return call_api("delete", uri, _.extend({
      prefix: prefix
    }, only(options, "keep_original")), callback, options);
  };

  exports.delete_resources_by_tag = function(tag, callback, options) {
    var resource_type, uri, _ref;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    uri = ["resources", resource_type, "tags", tag];
    return call_api("delete", uri, only(options, "keep_original"), callback, options);
  };

  exports.delete_derived_resources = function(derived_resource_ids, callback, options) {
    var uri;
    if (options == null) {
      options = {};
    }
    uri = ["derived_resources"];
    return call_api("delete", uri, {
      "derived_resource_ids": utils.build_array(derived_resource_ids)
    }, callback, options);
  };

  exports.tags = function(callback, options) {
    var resource_type, uri, _ref;
    if (options == null) {
      options = {};
    }
    resource_type = (_ref = options["resource_type"]) != null ? _ref : "image";
    uri = ["tags", resource_type];
    return call_api("get", uri, only(options, "next_cursor", "max_results", "prefix"), callback, options);
  };

  exports.transformations = function(callback, options) {
    if (options == null) {
      options = {};
    }
    return call_api("get", ["transformations"], only(options, "next_cursor", "max_results"), callback, options);
  };

  exports.transformation = function(transformation, callback, options) {
    var uri;
    if (options == null) {
      options = {};
    }
    uri = ["transformations", transformation_string(transformation)];
    return call_api("get", uri, only(options, "max_results"), callback, options);
  };

  exports.delete_transformation = function(transformation, callback, options) {
    var uri;
    if (options == null) {
      options = {};
    }
    uri = ["transformations", transformation_string(transformation)];
    return call_api("delete", uri, {}, callback, options);
  };

  exports.update_transformation = function(transformation, updates, callback, options) {
    var params, uri;
    if (options == null) {
      options = {};
    }
    uri = ["transformations", transformation_string(transformation)];
    params = only(updates, "allowed_for_strict");
    if (updates.unsafe_update != null) {
      params.unsafe_update = transformation_string(updates.unsafe_update);
    }
    return call_api("put", uri, params, callback, options);
  };

  exports.create_transformation = function(name, definition, callback, options) {
    var uri;
    if (options == null) {
      options = {};
    }
    uri = ["transformations", name];
    return call_api("post", uri, {
      transformation: transformation_string(definition)
    }, callback, options);
  };

  call_api = function(method, uri, params, callback, options) {
    var api_key, api_secret, api_url, cloud_name, cloudinary, xhr, _ref, _ref1, _ref2, _ref3, _ref4, _ref5;
    cloudinary = (_ref = (_ref1 = options["upload_prefix"]) != null ? _ref1 : config().upload_prefix) != null ? _ref : "https://api.cloudinary.com";
    cloud_name = (function() {
      var _ref3;
      if ((_ref2 = (_ref3 = options["cloud_name"]) != null ? _ref3 : config().cloud_name) != null) {
        return _ref2;
      } else {
        throw "Must supply cloud_name";
      }
    })();
    api_key = (function() {
      var _ref4;
      if ((_ref3 = (_ref4 = options["api_key"]) != null ? _ref4 : config().api_key) != null) {
        return _ref3;
      } else {
        throw "Must supply api_key";
      }
    })();
    api_secret = (function() {
      var _ref5;
      if ((_ref4 = (_ref5 = options["api_secret"]) != null ? _ref5 : config().api_secret) != null) {
        return _ref4;
      } else {
        throw "Must supply api_secret";
      }
    })();
    api_url = [cloudinary, "v1_1", cloud_name].concat(uri).join("/");
    if (method === "get") {
      api_url += "?" + utils.querystring.stringify(params);
    }
    xhr = Ti.Network.createHTTPClient({
      onload: function() {
        var result;
        if (_.include([200, 400, 401, 500], this.status)) {
          try {
            result = JSON.parse(this.responseText);
          } catch (e) {
            result = {
              error: {
                message: "Server return invalid JSON response. Status Code " + this.status
              }
            };
          }
          if (result["error"]) {
            result["error"]["http_code"] = this.status;
          } else {
            result["rate_limit_allowed"] = parseInt(this.getResponseHeader("x-featureratelimit-limit"));
            result["rate_limit_reset_at"] = new Date(this.getResponseHeader("x-featureratelimit-reset"));
            result["rate_limit_remaining"] = parseInt(this.getResponseHeader("x-featureratelimit-remaining"));
          }
          return callback(result);
        } else {
          return callback({
            error: {
              message: "Server returned unexpected status code - " + this.status
            }
          });
        }
      },
      onerror: function(e) {
        return callback({
          error: e
        });
      },
      timeout: (_ref5 = options["timeout"]) != null ? _ref5 : 60 * 1000
    });
    xhr.setRequestHeader('Authorization', 'Basic ' + Ti.Utils.base64encode(api_key + ':' + api_secret));
    xhr.open(method.toUpperCase(), api_url);
    xhr.setRequestHeader("Content-Type", "application/json; charset=utf-8");
    Ti.API.log('API XHR' + JSON.stringify(xhr));
    Ti.API.log('API params' + JSON.stringify(params));
    return xhr.send(JSON.stringify(params));
  };

  only = function() {
    var hash, key, keys, result, _i, _len;
    hash = arguments[0], keys = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
    result = {};
    for (_i = 0, _len = keys.length; _i < _len; _i++) {
      key = keys[_i];
      if (hash[key] != null) {
        result[key] = hash[key];
      }
    }
    return result;
  };

  transformation_string = function(transformation) {
    if (_.isString(transformation)) {
      return transformation;
    } else {
      return utils.generate_transformation_string(_.extend(transformation));
    }
  };

}).call(this);
